src/core/server/elasticsearch/client/cluster_client.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Client } from '@elastic/elasticsearch';
+import { getClientFacade } from './client_wrapper';
+import { ClientFacade } from './client_facade';
+import { configureClient } from './configure_client';
+import { Logger } from '../../logging';
+import { GetAuthHeaders, isRealRequest } from '../../http';
+import { Headers } from '../../http/router';
+import { ElasticsearchClientConfig } from './client_config';
+import { ScopedClusterClient, IScopedClusterClient } from './scoped_cluster_client';
+import { ScopeableRequest } from './types';
+import { ensureRawRequest, filterHeaders } from '../../http/router';
+
+const noop = () => undefined;
+
+interface IClusterClient {
+  asInternalUser: () => ClientFacade;
+  asScoped: (request: ScopeableRequest) => IScopedClusterClient;
+}
+
+export class ClusterClient implements IClusterClient {
+  private readonly internalWrapper: ClientFacade;
+  private readonly scopedClient: Client;
+
+  constructor(
+    private readonly config: ElasticsearchClientConfig,
+    logger: Logger,
+    private readonly getAuthHeaders: GetAuthHeaders = noop
+  ) {
+    this.internalWrapper = getClientFacade(configureClient(config, { logger }));
+    this.scopedClient = configureClient(config, { logger, scoped: true });
+  }
+
+  asInternalUser() {
+    return this.internalWrapper;
+  }
+
+  asScoped(request: ScopeableRequest) {
+    const headers = this.getScopedHeaders(request);
+    const scopedWrapper = getClientFacade(this.scopedClient, headers);
+    return new ScopedClusterClient(this.internalWrapper, scopedWrapper);
+  }
```

Comment:
Doing the same thing as in the legacy client by keeping a single reference of a scoped client to avoid parsing the config on each call of `asScoped`.

https://github.com/elastic/kibana/blob/56041f03ada9730e701508dbc10032b69de9f218/src/core/server/elasticsearch/cluster_client.ts#L197-L211

Reply:
hm...but the current version does this work on *every instantiation*. Doesn't it? we need to move https://github.com/elastic/kibana/pull/69905/files/175e0cb902b6235476f7d178a01d021960503a68#diff-735c304f3a374c21bed52dedab88917aR49 to this method.

Reply:
> hm...but the current version does this work on every instantiation. Doesn't it?

No it doesn't:

- The scoped `Client` is created once during the `ClusterClient` construction.
- The scoped `ClientWrapper` using the scoped client is instanciated once during each call to `asScoped(request)`, but uses the unique instance of the scoped `Client` (parsing the config is done during client, not wrapper, construction)